File: java/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<Integer>();
        Queue<TreeNode> queue = new LinkedList();
        if (root == null)
            return ans;
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size-- > 0) {
                TreeNode cur = queue.poll();
                if (size == 0)
                    ans.add(cur.val);
                if (cur.left != null)
                    queue.offer(cur.left);
                if (cur.right != null) 
                    queue.offer(cur.right);
            }
        }
        return ans;
    }
}

File: cpp/binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        rightView(root, ans, 0);
        return ans;
    }
    
private:
    void rightView(TreeNode* cur, vector<int>& ans, int curDepth) {
        if (!cur)
            return;
        if (ans.size() == curDepth)
            ans.push_back(cur -> val);
        rightView(cur -> right, ans, curDepth + 1);
        rightView(cur -> left, ans, curDepth + 1);
    }
};
